src/batch/src/executor/join/lookup_join.rs

```diff
@@ -257,95 +266,147 @@ impl<P: 'static + ProbeSideSourceBuilder> LookupJoinExecutor<P> {
+                    .collect_vec()
+            });
 
-                let probe_child = self.probe_side_source.build_source(&cur_row).await?;
-                let mut probe_side_stream = probe_child.execute();
-
-                let mut chunk_added = false;
-
-                // The handling of the different join types and which operation they should perform
-                // under each condition is very complex. The gist of it is:
-                //
-                // Inner Join: Always do_inner_join until probe_side_chunk is none.
-                //
-                // Left Outer Join: Always do_inner_join until probe_side_chunk is None. If this
-                //      occurs and no chunk has been added yet, null-pad the current row and add it.
-                //
-                // Left Semi Join: Always do_left_semi_anti_join until probe_side_chunk is None or
-                //      a chunk has been added, whichever is first.
-                //
-                // Left Anti Join: Always do_left_semi_anti_join until probe_side_chunk is None or
-                //      current row has a match on probe side, whichever is first. If the former
-                //      occurs and no match has been found yet, then add current row.
-                // TODO: Simplify the logic of handling the different join types
-                loop {
-                    let probe_side_chunk = probe_side_stream.next().await;
-                    let chunk_is_some = probe_side_chunk.is_some();
-
-                    let join_result = if let Some(chunk) = probe_side_chunk {
-                        let chunk = chunk?;
-
-                        match self.join_type {
-                            JoinType::Inner | JoinType::LeftOuter => {
-                                self.do_inner_join(&cur_row, chunk)
-                            }
-                            JoinType::LeftSemi | JoinType::LeftAnti if !chunk_added => {
-                                self.do_left_semi_anti_join(&cur_row, chunk)
+            let mut row_keys = vec![];
+            let mut all_row_refs = vec![];
+
+            groups.into_iter().for_each(|(row_key, row_refs)| {
+                row_keys.push(row_key);
+                all_row_refs.push(row_refs);
+            });
+
+            assert_eq!(row_keys.len(), all_row_refs.len());
+
+            self.probe_side_source.reset();
+            for row_key in &row_keys {
+                self.probe_side_source.add_scan_range(row_key)?;
+            }
+
+            let probe_child = self.probe_side_source.build_source().await?;
+            let mut probe_side_stream = probe_child.execute();
+
+            let mut probe_side_chunk_exists = true;
+
+            // for each row in the build side chunk, has_match tells us if it has a match
+            // on the probe side table
+            let mut has_match = all_row_refs
+                .iter()
+                .map(|rows| vec![false; rows.len()])
+                .collect_vec();
```

Comment:
So now you basically send a request for the whole build side chunk? 🤔

Reply:
Yes